hashes/src/chacha20poly1305/chacha20.rs

```diff
@@ -0,0 +1,380 @@
+        self.block(block);
+        keystream_at_slice(self.key, self.nonce, self.block_count, self.seek_offset_bytes)
+    }
+
+    /// Update the index of the keystream to the given byte.
+    pub fn seek(&mut self, seek: u32) {
+        self.block_count = seek / 64;
+        self.seek_offset_bytes = (seek % 64) as usize;
+    }
+
+    /// Update the index of the keystream to a block.
+    pub fn block(&mut self, block: u32) {
+        self.block_count = block;
+        self.seek_offset_bytes = 0;
+    }
+}
+
+fn quarter_round(state: &mut [u32; 16], a: usize, b: usize, c: usize, d: usize) {
+    state[a] = state[a].wrapping_add(state[b]);
+    state[d] = (state[d] ^ state[a]).rotate_left(16);
+    state[c] = state[c].wrapping_add(state[d]);
+    state[b] = (state[b] ^ state[c]).rotate_left(12);
+    state[a] = state[a].wrapping_add(state[b]);
+    state[d] = (state[d] ^ state[a]).rotate_left(8);
+    state[c] = state[c].wrapping_add(state[d]);
+    state[b] = (state[b] ^ state[c]).rotate_left(7);
+}
+
+fn double_round(state: &mut [u32; 16]) {
+    for (a, b, c, d) in CHACHA_ROUND_INDICIES {
+        quarter_round(state, a, b, c, d);
+    }
+}
+
+fn chacha_block(state: &mut [u32; 16]) {
+    let initial_state = *state;
+    for _ in 0..10 {
+        double_round(state)
+    }
+    for (modified, initial) in state.iter_mut().zip(initial_state.iter()) {
+        *modified = modified.wrapping_add(*initial)
+    }
+}
+
+fn prepare_state(key: [u8; 32], nonce: [u8; 12], count: u32) -> [u32; 16] {
+    let mut state: [u32; 16] = [0; 16];
+    state[0] = WORD_1;
+    state[1] = WORD_2;
+    state[2] = WORD_3;
+    state[3] = WORD_4;
+    state[4] =
+        u32::from_le_bytes(key[0..4].try_into().expect("infalliable conversion to 4 byte array"));
+    state[5] =
+        u32::from_le_bytes(key[4..8].try_into().expect("infalliable conversion to 4 byte array"));
+    state[6] =
+        u32::from_le_bytes(key[8..12].try_into().expect("infalliable conversion to 4 byte array"));
+    state[7] =
+        u32::from_le_bytes(key[12..16].try_into().expect("infalliable conversion to 4 byte array"));
+    state[8] =
+        u32::from_le_bytes(key[16..20].try_into().expect("infalliable conversion to 4 byte array"));
+    state[9] =
+        u32::from_le_bytes(key[20..24].try_into().expect("infalliable conversion to 4 byte array"));
+    state[10] =
+        u32::from_le_bytes(key[24..28].try_into().expect("infalliable conversion to 4 byte array"));
+    state[11] =
+        u32::from_le_bytes(key[28..32].try_into().expect("infalliable conversion to 4 byte array"));
+    state[12] = count;
+    state[13] =
+        u32::from_le_bytes(nonce[0..4].try_into().expect("infalliable conversion to 4 byte array"));
+    state[14] =
+        u32::from_le_bytes(nonce[4..8].try_into().expect("infalliable conversion to 4 byte array"));
+    state[15] = u32::from_le_bytes(
+        nonce[8..12].try_into().expect("infalliable conversion to 4 byte array"),
+    );
```

Comment:
Why not this?

```rust
for (state, word) in state[4..12].iter_mut().zip(key.chunks_exact(4)) {
    *state = u32::from_le_bytes(word.try_into().expect("chunks exact returns 4-byte slices");
}
```

Or even have `must state_iter = state.iter_mut()` and fill it with `*state_mut.next().unwrap() = WORD_1;` etc, pass the reference to for and thencontinue with the rest.

Reply:
Makes sense, I'll experiment with some options to condense this. We were trying to be as explicit as possible, but I can see how it might introduce bugs in this case.